src/group/libp2p_group_gossip_server.erl

```diff
@@ -1,105 +1,214 @@
+    PeerBookCount = get_opt(Opts, peerbook_connections, ?DEFAULT_PEERBOOK_CONNECTIONS),
+    InboundCount = get_opt(Opts, inbound_connections, ?DEFAULT_MAX_INBOUND_CONNECTIONS),
+    DropTimeOut = get_opt(Opts, drop_timeout, ?DEFAULT_DROP_TIMEOUT),
+    SeedNodes = get_opt(Opts, seed_nodes, []),
     self() ! {start_workers, PeerBookCount, length(SeedNodes)},
-    {ok, update_metadata(#state{sup=Sup, tid=TID, client_spec=ClientSpec,
-                                seed_nodes=SeedNodes, peerbook_connections=PeerBookCount,
-                                drop_timeout=DropTimeOut, drop_timer=schedule_drop_timer(DropTimeOut)})}.
+    {ok, update_metadata(#state{sup=Sup, tid=TID,
+                                seed_nodes=SeedNodes,
+                                max_inbound_connections=InboundCount,
+                                peerbook_connections=PeerBookCount,
+                                drop_timeout=DropTimeOut,
+                                drop_timer=schedule_drop_timer(DropTimeOut)})}.
+
+handle_call({accept_stream, _Session, _StreamPid}, _From, State=#state{workers=[]}) ->
+    {reply, {error, not_ready}, State};
+handle_call({accept_stream, Session, StreamPid}, From, State=#state{}) ->
+    libp2p_session:identify(Session, self(), {From, StreamPid}),
+    {noreply, State};
+handle_call({handle_data, Key, Msg}, _From, State=#state{}) ->
+    %% Incoming message from a gossip stream for a given key
+    case maps:find(Key, State#state.handlers) of
+        error -> {reply, ok, State};
+        {ok, {M, S}} -> case (catch M:handle_gossip_data(Msg, S)) of
+                               {'EXIT', Err} -> {reply, {error, Err}, State};
+                               {error, Reason} -> {reply, {error, Reason}, State};
+                               _ -> {reply, ok, State}
+                           end
+    end;
+handle_call({connected_addrs, Kind}, _From, State=#state{}) ->
+    {Addrs, _Pids} = lists:unzip(connections(Kind, State)),
+    {reply, Addrs, State};
 
-handle_call(workers, _From, State=#state{}) ->
-    {reply, connections(all, State), State};
 handle_call(Msg, _From, State) ->
     lager:warning("Unhandled call: ~p", [Msg]),
     {reply, ok, State}.
 
+handle_cast({add_handler, Key, Handler}, State=#state{handlers=Handlers}) ->
+    {noreply, State#state{handlers=maps:put(Key, Handler, Handlers)}};
+handle_cast({remove_handler, Key}, State=#state{handlers=Handlers}) ->
+    {noreply, State#state{handlers=maps:remove(Key, Handlers)}};
+
+handle_cast({request_target, inbound, WorkerPid}, State=#state{}) ->
+    {noreply, stop_inbound_worker(WorkerPid, State)};
 handle_cast({request_target, Kind=peerbook, WorkerPid}, State=#state{tid=TID}) ->
-    PeerAddrs = [ libp2p_crypto:address_to_p2p(Key)
-                  || Key <- libp2p_peerbook:keys(libp2p_swarm:peerbook(TID)) ],
+    PeerKeys = try
+                   libp2p_peerbook:keys(libp2p_swarm:peerbook(TID))
+               catch
```

Comment:
Why would this fail?

Reply:
During shutdown the gossip_server could try and reach the peerbook. which causes a loud crash

Reply:
I'll add a comment to that effect